lib/ocpp/v201/charge_point.cpp

```diff
@@ -3296,6 +3324,145 @@ void ChargePoint::handle_heartbeat_response(CallResult<HeartbeatResponse> call)
+        reservation_available = false;
+        status_info = "Reservation is not enabled";
+    }
+
+    if (!reservation_available) {
+        // Reservation not available / implemented, return 'Rejected'.
+        // H01.FR.01
+        EVLOG_info << "Receiving a reservation request, but reservation is not enabled or implemented.";
+        send_reserve_now_rejected_response(call.uniqueId, status_info);
+        return;
+    }
+
+    // Check if we need a specific evse id during a reservation and if that is the case, if we recevied an evse id.
+    const ReserveNowRequest request = call.msg;
+    if (!request.evseId.has_value() &&
+        !this->device_model->get_optional_value<bool>(ControllerComponentVariables::ReservationCtrlrNonEvseSpecific)
+             .value_or(false)) {
+        // H01.FR.19
+        EVLOG_warning << "Trying to make a reservation, but no evse id was given while it should be sent in the "
+                         "request when NonEvseSpecific is disabled.";
+        send_reserve_now_rejected_response(
+            call.uniqueId,
+            "No evse id was given while it should be sent in the request when NonEvseSpecific is disabled");
+        return;
+    }
+
+    const std::optional<int32_t> evse_id = request.evseId;
+
+    if (evse_id.has_value()) {
+        if (!evse_manager->does_evse_exist(evse_id.value())) {
+            EVLOG_error << "Trying to make a reservation, but evse " << evse_id.value() << " is not a valid evse id.";
+            send_reserve_now_rejected_response(call.uniqueId, "Evse id does not exist");
+            return;
+        }
+
+        // Check if there is a connector available for this evse id.
+        std::optional<ConnectorStatusEnum> status =
+            get_connector_status(static_cast<uint32_t>(evse_id.value()), request.connectorType);
+
+        if (!status.has_value()) {
+            EVLOG_info << "Trying to make a reservation for connector type "
+                       << conversions::connector_enum_to_string(request.connectorType.value_or(ConnectorEnum::Unknown))
+                       << " for evse " << evse_id.value() << ", but this connector type does not exist.";
+            send_reserve_now_rejected_response(call.uniqueId, "Connector type does not exist");
+            return;
+        }
+    } else {
+        // No evse id. Just search for all evse's if there is something available for reservation
+        const uint64_t number_of_evses = evse_manager->get_number_of_evses();
+        if (number_of_evses <= 0) {
+            send_reserve_now_rejected_response(call.uniqueId, "No evse's found in charging station");
+            EVLOG_error << "Trying to make a reservation, but number of evse's is 0";
+            return;
+        }
+
+        bool status_found = false;
+        for (uint64_t i = 1; i <= number_of_evses; i++) {
+            std::optional<ConnectorStatusEnum> status = get_connector_status(i, request.connectorType);
```

Comment:
use `is_connector_available` here

Reply:
But then it can also be occupied or reserved.